File: docs/core/examples/ruby/real_time_transaction_processing.rb
require "chain"

chain = Chain::Client.new
$chain = chain # global alias for thread funcs
signer = Chain::HSMSigner.new

key = chain.mock_hsm.keys.create
signer.add_key(key, chain.mock_hsm.signer_conn)

chain.assets.create(alias: 'gold', root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'alice', root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'bob', root_xpubs: [key.xpub], quorum: 1)

# snippet create-feed
chain.transaction_feeds.create(
  alias: 'local-transactions',
  filter: "is_local='yes'"
)
# endsnippet

def run_processing_loop
  chain = $chain

  # snippet get-feed
  feed = chain.transaction_feeds.get(alias: 'local-transactions')
  # endsnippet

  # snippet processing-loop
  feed.consume do |tx|
    process_transaction(tx)
    feed.ack
  end
  # endsnippet
end

# snippet processor-method
def process_transaction(tx)
  puts "New transaction at #{tx.timestamp}"
  puts "\tID: #{tx.id}"

  tx.inputs.each_with_index do |input, index|
    puts "\tInput #{index}"
    puts "\t\tType: #{input.type}"
    puts "\t\tAsset: #{input.asset_alias}"
    puts "\t\tAmount: #{input.amount}"
    puts "\t\tAccount: #{input.account_alias}"
  end

  tx.outputs.each_with_index do |output, index|
    puts "\tOutput #{index}"
    puts "\t\tType: #{output.type}"
    puts "\t\tPurpose: #{output.purpose}"
    puts "\t\tAsset: #{output.asset_alias}"
    puts "\t\tAmount: #{output.amount}"
    puts "\t\tAccount: #{output.account_alias}"
  end
end
# endsnippet

# snippet processing-thread
Thread.new { run_processing_loop }
# endsnippet

# snippet issue
issuance = chain.transactions.build do |b|
  b.issue asset_alias: 'gold', amount: 100
  b.control_with_account account_alias: 'alice', asset_alias: 'gold', amount: 100
end

chain.transactions.submit(signer.sign(issuance))
# endsnippet

sleep(0.25)

# snippet transfer
transfer = chain.transactions.build do |b|
  b.spend_from_account account_alias: 'alice', asset_alias: 'gold', amount: 50
  b.control_with_account account_alias: 'bob', asset_alias: 'gold', amount: 50
end

chain.transactions.submit(signer.sign(transfer))
# endsnippet

sleep(0.25)
exit

File: docs/core/examples/node/realTimeTransactionProcessing.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let key, feed

// snippet processor-method
processTransaction = (tx) => {
  console.log(`New transaction at ${tx.timestamp}`)
  console.log(`\tID: ${tx.id}`)

  tx.inputs.forEach((input, index) => {
    console.log(`\tInput ${index}`)
    console.log(`\t\tType: ${input.type}`)
    console.log(`\t\tAsset: ${input.asset_alias}`)
    console.log(`\t\tAmount: ${input.amount}`)
    console.log(`\t\tAccount: ${input.account_alias}`)
  })

  tx.outputs.forEach((output, index) => {
    console.log(`\tOutput ${index}`)
    console.log(`\t\tType: ${output.type}`)
    console.log(`\t\tPurpose: ${output.purpose}`)
    console.log(`\t\tAsset: ${output.asset_alias}`)
    console.log(`\t\tAmount: ${output.amount}`)
    console.log(`\t\tAccount: ${output.account_alias}`)
  })
}
// endsnippet

client.mockHsm.keys.create()
.then(_key => {
  key = _key.xpub
  signer.addKey(key, client.mockHsm.signerUrl)
}).then(() => Promise.all([
  client.assets.create({
    alias: 'gold',
    root_xpubs: [key],
    quorum: 1,
  }),
  client.accounts.create({
    alias: 'alice',
    root_xpubs: [key],
    quorum: 1
  }),
  client.accounts.create({
    alias: 'bob',
    root_xpubs: [key],
    quorum: 1
  }),

  // snippet create-feed
  client.transactionFeeds.create({
    alias: 'local-transactions',
    filter: "is_local='yes'"
  })
  // endsnippet
]))
.then(() => {

  // snippet get-feed
  const feedPromise = client.transactionFeeds.get({
    alias: 'local-transactions'
  })
  // endsnippet

  return feedPromise
})
.then(feed => {

  // snippet processing-loop
  const processingLoop = (tx, next, done, fail) => {
    processTransaction(tx)
    next(true)
  }
  // endsnippet

  // snippet processing-thread
  // JavaScript is single-threaded, and uses a callback for processing
  feed.consume(processingLoop)
  // endsnippet
})
.then(() =>
  // snippet issue
  client.transactions.build( function(builder){
    builder.issue({
      asset_alias: 'gold',
      amount: 100
    })
    builder.controlWithAccount({
      account_alias: 'alice',
      asset_alias: 'gold',
      amount: 100
    })
  })
  .then((issuance) => signer.sign(issuance))
  .then((signed) => client.transactions.submit(signed))
  // endsnippet
).then(() =>
  // snippet transfer
  client.transactions.build( function(builder){
    builder.spendFromAccount({
      account_alias: 'alice',
      asset_alias: 'gold',
      amount: 50
    })
    builder.controlWithAccount({
      account_alias: 'bob',
      asset_alias: 'gold',
      amount: 50
    })
  })
  .then((issuance) => signer.sign(issuance))
  .then((signed) => client.transactions.submit(signed))
  // endsnippet
)
